src/typechecker.erl

```diff
@@ -1568,7 +1568,10 @@ type_check_logic_op(Env, Op, P, Arg1, Arg2) ->
             throw({type_error, boolop, Op, P, Ty1});
         {true, Cs2} ->
             {Ty2, VB2, Cs3} = type_check_expr(Env#env{ venv = UnionVarBindsSecondArg(Env#env.venv,VB1 )}, Arg2),
-            case subtype(Ty2, {type, P, bool, []}, Env#env.tenv) of
+            % Allow any() in second argument for shortcut operators
+            SndArgTy = if Op == 'andalso'; Op == 'orelse' -> type(any);
+                          true                            -> type(bool) end,
+            case subtype(Ty2, SndArgTy, Env#env.tenv) of
```

Comment:
nitpicking for below unchanged line `{normalize({type, erl_anno:new(0), union, [Ty1, Ty2]}, Env#env.tenv)`

I think the inferred type of short-circuit ops should not be `union(Ty1, Ty2)` but `union(false, Ty2)` or `union(true, Ty2)` (respectively for andalso and orelse - as you stated in the typing rule)

because the below code would trigger a type error
```
f() ->
  B = get_bool(),
  I = get_int(),
  R = B andalso I,
  g(R).

-spec get_bool() -> boolean().
-spec get_int() -> integer().
-spec g(false | integer()) -> any().
```
`g/1` is called with `R` which has the inferred type of `boolean() | integer()`, which is not a subtype of expected argument type `false | integer()`

This is a different code path than type checking, where we make the outer result type flow down/inwards/backwards.

Reply:
Well spotted, I'll update it.

Reply:
Also the `union(Ty1, Ty2)` is wrong for `xor`. This will infer `true xor true :: true`.

Reply:
Pushed a fix. I relaxed the inferred type to `boolean()` for non-shortcut ops. While we could technically be more precise, I don't think it's worth it.